### PR TITLE
feat: Implement Toast using composition API

### LINKED ISSUE #27773: Implement Toast component
Related to https://github.com/microsoft/fluentui/issues/26638

Fully implement the Toast component that behaves like the temporary one introduced in #27772

### PR DESCRIPTION
**Pull request by ling1726**
Implements the `Toast` component using the composition API. Also updates the toast imperative functions to be able to override the root slot.

Fixes #27773

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/3d73007c50ca4008618498a998cdaed5675bbda6'>3d73007c50ca4008618498a998cdaed5675bbda6</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount_base.html">10</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount.html">18</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Avatar-mount_base.html">594</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Avatar-mount.html">583</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Button-mount_base.html">299</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Button-mount.html">306</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Field-mount_base.html">1062</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Field-mount.html">1059</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProvider-mount_base.html">643</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProvider-mount.html">653</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-mount_base.html">85</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-mount.html">81</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">77</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">69</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">71</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount_base.html">10</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount.html">18</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/MakeStyles-mount_base.html">841</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/MakeStyles-mount.html">845</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Persona-mount_base.html">1598</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/Persona-mount.html">1579</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/SpinButton-mount_base.html">1240</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28094/perf-test-react-components/SpinButton-mount.html">1269</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

File: packages/react-components/react-toast/src/components/Timer/Timer.tsx [hunk 1]
```diff
@@ -27,5 +21,15 @@ export const Timer: React.FC<{
...
+      onAnimationEnd={onTimeout}
+      data-timer-status={style.animationPlayState}
+      ref={ref}
+      style={style}
+      className={styles.progress}
+    />
+  );
+});
+
+Timer.displayName = 'Timer';
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Timer/Timer.tsx:35]: ToastTimer?

File: packages/react-components/react-toast/src/components/Toast/Toast.types.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+import { Announce } from '../AriaLive/AriaLive.types';
+import { Toast as VanillaToast } from '../../state';
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toast/Toast.types.ts:4]: This type collision does not seem like something good. Did you consider renaming it? ToastItem?

File: packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts [hunk 3]
```diff
@@ -1,19 +1,26 @@
...
 import { tokens } from '@fluentui/react-theme';
+import type { ToastSlots, ToastState } from './Toast.types';
+import type { SlotClassNames } from '@fluentui/react-utilities';
 
-export const useToastStyles = makeStyles({
-  toast: {
-    pointerEvents: 'all',
+export const toastClassNames: SlotClassNames<ToastSlots> = {
+  root: 'fui-Toast',
+  timer: 'fui-Timer',
```
Comment (layershifter) [inline packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts:8]: fui-ToastTimer

File: packages/react-components/react-toast/src/components/Toast/Toast.types.ts (hunk 2 above)
Comment (ling1726) [reply packages/react-components/react-toast/src/components/Toast/Toast.types.ts:4]: yeah this should be addressed, but should be done in a PR scoped to this https://github.com/microsoft/fluentui/issues/28108

File: packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts (hunk 3 above)
Comment (layershifter) [reply packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts:8]: BTW, is it used at all?

**Review by layershifter — APPROVED** [review]

File: packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts (hunk 3 above)
Comment (ling1726) [reply packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts:8]: no, but this component in general is not public, I'm just following our slot architecture

Comment (layershifter) [reply packages/react-components/react-toast/src/components/Toast/useToastStyles.styles.ts:8]: I see, then at least please rename the class as it's a slot of a component

**Pull request merged**